iree/compiler/Dialect/Stream/Transforms/ScheduleConcurrency.cpp

```diff
@@ -0,0 +1,281 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Stream/Analysis/Partitioning.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamDialect.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamOps.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamTypes.h"
+#include "iree/compiler/Dialect/Stream/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Stream/Transforms/Passes.h"
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilTypes.h"
+#include "llvm/ADT/BitVector.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+
+#define DEBUG_TYPE "iree-stream-schedule-concurrency"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Stream {
+namespace {
+
+// TODO(benvanik): deduplicate this with ScheduleExecution - almost all of this
+// is identical.
+
+// Incremental builder for a partitioned region of executable work.
+// Must be constructed in a topological order of all partitions.
+struct WavePartitionBuilder {
+  explicit WavePartitionBuilder(Block *parentBlock, size_t ordinal,
+                                Partition *partition,
+                                BlockAndValueMapping &parentMapping,
+                                MLIRContext *context)
+      : ordinal(ordinal), partition(partition), builder(context) {
+    // Fuse the location of all ops we'll be putting in the partition.
+    SmallVector<Location> locs;
+    for (auto *op : partition->ops) {
+      locs.push_back(op->getLoc());
+    }
+    auto fusedLoc = FusedLoc::get(context, locs);
+
+    // Find the insertion point in the parent block.
+    // This is at the last op defining an input as all inputs must be available.
+    Operation *insertionPt = nullptr;
+    for (auto in : partition->ins) {
+      auto *definingOp = in.getDefiningOp();
+      if (!definingOp) continue;
+      if (definingOp->getBlock() != parentBlock) continue;
+      if (!insertionPt) {
+        insertionPt = definingOp;
+      } else if (insertionPt->isBeforeInBlock(definingOp)) {
+        insertionPt = definingOp;
+      }
```

Comment:
(similar to another comment)
intentional that the `if` and `else if` hae the same body?